<center>
  <a href="MLSD-04-FeatureEngineering-B.ipynb" target="_self">Feature Engineering B</a> | <a href="./">Content Page</a> | <a href="MLSD-04-FeatureEngineering-Ex-2.ipynb">Feature Engineering Exercise 2</a>
</center>

# <center>FEATURE ENGINEERING: EXERCISE 1</center>

<center><b>Copyright &copy 2023 by DR DANNY POO</b><br> e:dannypoo@nus.edu.sg<br> w:drdannypoo.com</center><br>

# Tasks
- To read in and explore data set.
- To produce <b>missing value indicator</b>.
- To carry out <b>Median Data Imputation</b>.
- To carry out <b>one-hot encoding</b>.

## Read in and Explore Data Set

In [54]:
# Import libraries
# Write code here

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme()
sns.set_palette(sns.color_palette(['#851836', '#edbd17']))
sns.set_style("darkgrid")

from sklearn.preprocessing import MinMaxScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split


In [55]:
# Read in data
df = pd.read_csv('./data/titanic/train.csv')
df[0:10]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [56]:
# Display mean value of number of NaN value in columns
# Write code here
df.isnull().mean()


PassengerId    0.000000
Survived       0.000000
Pclass         0.000000
Name           0.000000
Sex            0.000000
Age            0.198653
SibSp          0.000000
Parch          0.000000
Ticket         0.000000
Fare           0.000000
Cabin          0.771044
Embarked       0.002245
dtype: float64

In [16]:
# Examine data format
# Write code here

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


**Observations**
- There are missing values in Age, Cabin and Embarked columns.

In [17]:
# Check how many observations have NaN values
# Write code here

print(f"Total number of passengers with values in all variables: {df['PassengerId'].count()}")
print(f"Total number of passengers in the Titanic dataset: {df.notna().all(axis=1).sum()}")
print(f"Percentage of data without missing values: {183/891}")



Total number of passengers with values in all variables:  183
Total number of passengers in the Titanic dataset:  891
Percentage of data without missing values:  0.2053872053872054


**Observations**<br>
- There is only 20.5% of observations that have complete information.

## Missing Value Indicator

In [28]:
# Create new column 'Age_NA', column has 1 if the 'Age' column observation is missing, or 0 otherwise.
# Write code here

df['Age_NA'] = df['Age'].isnull().astype(int)


In [19]:
# Display first ten values of 'Age' and 'Age_NA' columns
# Write code here

df[['Age','Age_NA']].head(10)


Age  Age_NA
0  22.0       0
1  38.0       0
2  26.0       0
3  35.0       0
4  35.0       0
5   NaN       1
6  54.0       0
7   2.0       0
8  27.0       0
9  14.0       0

**Observations**<br>
- Age in row 6 is NaN.
- Age_NA variable is created to capture the missingness.

## Median Data Imputation

In [20]:
# Fill in median value of Age in Age variable where value is NaN
# Display the first ten rows of df
# Write code here

df['Age'] = df['Age'].fillna(df['Age'].median())

df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male  28.0      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  Age_NA  
0      0         A/5 21171   7.2500   NaN        S       0  
1      0          PC 17599  71.2833   C85        C       0  
2      0  STON/O2. 3101282   7.9250   NaN        S       0  
3      0            113803  53.1000  C123        S       0  
4      0            373450   8.0500   NaN        S       0  
5      0            330877   8.4583   NaN        Q       1  
6      0             17463  51.8625   E46        S       0  
7      1            349909  21.0750   NaN        S       0  
8      2            347742  11.1333   NaN        S       0  
9      0            237736  30.0708   NaN        C       0

## One-hot Encoding

In [44]:
# Display first 5 rows of Sex variable
# Write code here
new_df = pd.DataFrame(df['Sex'])
new_df.head(5)


Sex
0    male
1  female
2  female
3  female
4    male

In [41]:
# One-hot encode Sex variable and display first 5 rows of the two encoded columns
# Write code here
onehotenc_sex_df = pd.get_dummies(df['Sex'])

onehotenc_sex_df.head(5)



female  male
0       0     1
1       1     0
2       1     0
3       1     0
4       0     1

In [49]:
# Concatenate Sex variable with the newly created encoded columns and display as shown
# Write code here

new_concat_df = pd.concat([new_df, onehotenc_sex_df], axis=1)

new_concat_df

Sex  female  male
0      male       0     1
1    female       1     0
2    female       1     0
3    female       1     0
4      male       0     1
..      ...     ...   ...
886    male       0     1
887  female       1     0
888  female       1     0
889    male       0     1
890    male       0     1

[891 rows x 3 columns]

In [51]:
# Get n-1 dummy variables as shown. Hint: use drop_first=True parameter in get_dummies()
# Write code here

pd.get_dummies(df['Sex'], drop_first=True)


male
0       1
1       0
2       0
3       0
4       1
..    ...
886     1
887     0
888     0
889     1
890     1

[891 rows x 1 columns]

<center>
  <a href="MLSD-04-FeatureEngineering-B.ipynb" target="_self">Feature Engineering B</a> | <a href="./">Content Page</a> | <a href="MLSD-04-FeatureEngineering-Ex-2.ipynb">Feature Engineering Exercise 2</a>
</center>